rfcs/0008-storage-handler.md

```diff
@@ -0,0 +1,338 @@
+1. Dynamic storage types, such as `StorageMap` and `StorageVec`, currently require a "hacky" compiler intrinsic called `__get_storage_key` which is not very well defined and is hard to use. Introducing the concept of a `StorageHandler` will helps us completely remove this intrinsic.
+1. The current implementation of dynamic storage types prevents th-em from being used as struct fields or as type parameters of other storage types (e.g. `StorageVec<StorageVec<u64>>`).
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+A "storage handler" is a thin wrapper type, in the standard library, around a `b256` key that describes a storage slot:
+
+```rust
+pub struct StorageHandler<T> {
+    key: b256,
+}
+```
+
+The type `T` describes the type of the data stored at `key`. Since each storage slot is 64 bytes long, some types cannot fit in a single storage slot. In that case, the data is stored across multiple consecutive storage slots starting at `key`. More on this in [Storage Slot Assignment](#storage-slot-assignment).
+
+Reading and writing the data stored at `key` can be accomplished using the `get` and `store` methods below:
+
+```rust
+impl<T> StorageHandler<T> {
+    #[storage(read)]
+    pub fn get(self) -> Option<T> {
+        std::storage::get(self.key)
+    }
+
+    #[storage(write)]
+    pub fn store(self, other: T) {
+        std::storage::store(self.key, other);
+    }
+}
+```
+
+Notice that `get` returns an `Option` because it is possible that a particular storage slot is not valid, i.e. has not been written before, in which case `None` is returned. If the type `T` spans multiple storage slots, then the method `get` returns `None` if at least one the storage slots that `T` spans is invalid.
+
+With the `StorageHandler` type available, this RFC proposes a redefinition of the meaning of the expression `storage.<var>.<field>...` to return a `StorageHandler` "pointing" to the actual data instead of returning the data itself. The RFC also proposes removing the "reassignment" statement `storage.<var>.<field>.. = ..` from the language.
+
+Below is an example showing the behavior of `StorageHandler` and the new behavior of storage access expressions:
+
+```rust
+struct MyStruct {
+    x: u64,
+    y: b256,
+}
+
+storage {
+    s: MyStruct = MyStruct { x: 0, y: ZERO_B256 }, 
+}
+
+fn foo() {
+    // Get a storage handler to the struct field `x` directly and then use it to store `42` in `x`.
+    let x_handler: StorageHandler<u64> = storage.s.x;
+    x_handler.store(42);
+   
+    // Call `store` and `get` directly on `storage.s.y`
+    storage.s.y.store(ZERO_B256);
```

Comment:
Is there any benefit to allowing users to call `store` on `storage.s.y` directly here without using a `StorageHandler`?
It might be simpler to just allow the one way, i.e: 
```
let y_handler: StorageHandler<u64> = storage.s.y;
    y_handler.store(ZERO_B256);
```

Reply:
This really is just a byproduct of the fact that `storage.x.y` is an expression and so calling `.store` directly or through a temporary variable is intrinsically supported by the language, so no extra work there :)